internal/mount/mount.go

```diff
@@ -0,0 +1,141 @@
+package mount
+
+import (
+	"fmt"
+	"os"
+	"strings"
+	"sync"
+
+	"git.sr.ht/~spc/go-log"
+	"github.com/openshift/assisted-installer-agent/src/util"
+	"github.com/project-flotta/flotta-operator/models"
+	"golang.org/x/sys/unix"
+)
+
+const (
+	filesystemsFile = "/etc/filesystems"
+)
+
+type Manager struct {
+	lock sync.Mutex
+	dep  util.IDependencies
+
+	// filesystems holds the content of /etc/filesystems
+	// it used to validate the mount type
+	filesystems string
+}
+
+func New() (*Manager, error) {
+	content, err := os.ReadFile(filesystemsFile)
+	if err != nil {
+		log.Warnf("Cannot list content of '%s': %s", filesystemsFile, err)
+		return nil, fmt.Errorf("cannot list content of '%s': %s", filesystemsFile, err)
+	}
+
+	return &Manager{
+		filesystems: string(content),
+		dep:         util.NewDependencies("/"),
+	}, nil
+}
+
+func (m *Manager) Init(config models.DeviceConfigurationMessage) error {
+	return m.Update(config)
+}
+
+func (m *Manager) Update(config models.DeviceConfigurationMessage) error {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	_, currentMounts, err := GetMounts(m.dep)
+	if err != nil {
+		return err
+	}
+
+	// holds the directories mounted in the current _Update_ call.
+	// It helps avoiding mounting the same directory twice.
+	alreadyMounted := make(map[string]struct{})
+
+	for _, mm := range config.Configuration.Mounts {
+		if _, found := alreadyMounted[mm.Directory]; found {
+			log.Warnf("Directory '%s' has already been mounted. Skipping..", mm.Directory)
+			continue
+		}
+
+		if err := m.isValid(mm); err != nil {
+			log.Errorf("Mount configuration '%+v' not valid: %s", mm, err)
+			continue
+		}
+
+		// if the directory is mounted (not in the current call) and the configuration is different
+		// try to umount it first and than mount it with the new configuration.
+		if c, found := currentMounts[mm.Directory]; found {
+			if !isEqual(c, mm) {
+				if err := umount(c); err != nil {
```

Comment:
shouldn't retry here? not a hard requirement, will be fixed on the next upload, but yeah :\

Reply:
it will rety on its own on next heartbeat. The user asked for 2 mounts and one is falling because it is already mounted on a pod. As long as the user does not stop the pod to be able to unmount there will be a difference between what we report back to user and what he wants.
Actually, there is a subtle bug here cause we are reporting all the mounts and not only ones asked by the user so there will be always a difference.